### ISSUE TITLE
imagestream is always using the registry ip instead of registry hostname

### ISSUE THREAD

**Issue by thincal** [openshift/origin#6283]
If I push an image for example `registry.example.com:5000/default/hello-openshift` to the integrated registry, at the same time an imagestream will be created automatically, but `oc get is` shows:

```
NAME              DOCKER REPO                                   TAGS      UPDATED
hello-openshift   172.30.117.244:5000/default/hello-openshift   latest    7 minutes ago
```

When I create a pod referring the image `registry.example.com:5000/default/hello-openshift`, it will report the error:

```
Failed to pull image "registry.example.com:5000/default/hello-openshift:latest": image pull failed for registry.example.com:5000/default/hello-openshift:latest, this may be because there are no credentials on this request. details: (Error: image default/hello-openshift:latest not found) 
```

So what the reason to use the registry ip in the imagestream instead of the registry hostname even if I push the image formatted like as hostname:port/project/imagename ? or is this the bug simply ?

This issue is confusing me several days about the image pull failure when creating the pod, after using the registry-ip:port/project/imagename pod is created successfully.

**Comment by thincal** [comment]
And since the registry service's IP could be changed, so if referring the IP in the imagestream, is that good ?

for example:
```
$ oc get is
NAME                              DOCKER REPO                                                     TAGS                         UPDATED
hello-openshift                   172.30.117.244:5000/openshift/hello-openshift                   latest                       2 days ago
```

why not:
```
$ oc get is
NAME                              DOCKER REPO                                                     TAGS                         UPDATED
hello-openshift                   registry.example.com:5000/openshift/hello-openshift                   latest                       2 days ago
```

**Comment by thincal** [comment]
@soltysh so is this by design ?

**Comment by soltysh** [comment]
Unfortunately for now it is, but I'm going to work on it. I can't promise
when, though,   but I'll try as soon as possible.

**Comment by liggitt** [comment]
switching to hostname requires changes to the dockercfgs we generate before pulls will work. @deads2k

**Comment by deads2k** [comment]
> switching to hostname requires changes to the dockercfgs we generate before pulls will work. @deads2k

I'm looking at it.

**Comment by thincal** [comment]
@deads2k any comments so far with this question ?

**Comment by liggitt** [comment]
Changes to credential generation are in progress but won't be completed until the new year

**Comment by liggitt** [comment]
Once that is done, the image stream resolution could be changed to use DNS instead of IP. That would add a dns resolution requirement for services to the nodes. I'm not positive that is required today, so that may be a new requirement.

**Comment by thincal** [comment]
Thanks for your update.

> That would add a dns resolution requirement for services to the nodes.

I don't understand this quite well, since the registry hostname resolution to IP is done by external DNS server, what do you mean about here? thanks.

**Comment by liggitt** [comment]
There are multiple DNS names that could be used for the registry. The service DNS name (e.g. `docker-registry.default.svc.cluster.local`) only resolves using the in-cluster DNS. I meant that for the nodes to successfully pull using that name, they would have to be configured to use the in-cluster DNS nameserver, and I wasn't sure if they were or not. 

The second set of DNS names that could be used would be route host names.

**Comment by thincal** [comment]
@liggitt, any plan or progress about using the hostname for imagestream now?
sorry for pushing, but really looking forward to getting any news. thanks :)

**Comment by thincal** [comment]
@liggitt @soltysh @deads2k sorry for disturbing you, is there any news about this issue ? many thanks really :)

**Comment by liggitt** [comment]
Not yet. Still desired, but not at the top of the priority list.

**Comment by soltysh** [comment]
@pweil- since this is a repeating problem do you think we could address that during next planning? If @liggitt hasn't started working on it yet I'd be willing to pick it up during next sprint.

**Comment by deads2k** [comment]
@soltysh pull in progress to make it easier to use the route is here: https://github.com/openshift/origin/pull/6385 .  Its the plumbing we need to rebuild the secrets to enable the other refs to resolve.

After this we have to resolve the certificate problem.

**Comment by smarterclayton** [comment]
Using this as quick writeup:

1. It must be possible for an API client to determine the internal DNS name of the given repo (status.dockerImageRepository)
2. It must be possible for an API client to determine the external DNS name of the given repo (new field)
3. I don't think we need to support more than one of either, but we could introduce a new array field if necessary
4. We want to move the docker registry from `docker-registry.default.svc` to `registry.openshift.svc` if possible (this may require migration)
5. The external name should only be set if the registry is exposed
6. It must continue to be possible to override the internal registry name and the external registry name - we should switch to pulling that from config, not from environment
7. When we export, we should use the external name
8. The image policy admission controller should treat internal and external name as the same
9. Image migration may need updating

**Comment by smarterclayton** [comment]
@sdodson FYI

**Comment by mfojtik** [comment]
@miminar @legionus ^

**Comment by soltysh** [comment]
@mfojtik I remember you were working on something wrt this, weren't you?

**Comment by mfojtik** [comment]
@soltysh I think it is possible to set the OPENSHIFT_DEFAULT_REGISTRY to a service name today and use it instead of the IP address. Using external/public route for the DockerRepo is not done yet (there is a PR open to discuss a new API field with publicPullSpec).

**Comment by mfojtik** [comment]
Closing as this is >1y old and starting from 3.6 we default to registry service DNS.

**Comment by shrishs** [comment]
Today I have upgraded from v3.6.173.0.21 to v3.6.173.0.49 and I have seen the registry is pushing again using registry service ip and thereby failing as this ip is not there in NO_PROXY list.

**Comment by miminar** [comment]
@shrishs how do you install the cluster? Containerized or advanced installation with ansible? Did you set `OPENSHIFT_DEFAULT_REGISTRY` somewhere?

**Comment by cjsrinivas** [comment]
This is a problem, since a project export and restore, through oc export and oc create commands fail, because the image streams are pointing at the wrong service ip for the image stream. Is there a solution for this, since getting to move 7 production OSE instances to 3.6 is not an immediate option.

**Comment by miminar** [comment]
@cjsrinivas Exporting image streams is a complex topic with no perfect solution. Is the purpose of export a backup of the current state of current cluster? Or should the exported streams point to a new registry in another cluster with migrated images? Or should it be imported in other cluster with URLs pointing back to the original registry's external URL in order to use pullthrough to migrate the images?

I certainly haven't covered all the possible use-cases, maybe not event your use-case.

Currently, we don't provide export functionality generic enough to cover those. We'd certainly like to improve the situation though. If you desire such feature and/or have a constructive feedback/suggestion in this regard, please open a new issue, where we can track that.

**Comment by cjsrinivas** [comment]
The use case is fairly simple, following the instructions on docs.openshift.com to backup and restore projects. OpenShift instance #1 is up and running, we run a sequence of oc commands to export various objects to YAML files. We use these YAML files to restore the project on OpenShift instance #2. Since OpenShift #2 is different, the IP address for the internal docker registry is different, the image streams do not work. Question here is, if both instances used the internal docker registry, do you see any problems when the project is restored if the image stream's ip address is replaced with the internal host name of the service in the backup files. In other words, will this work?

**Comment by boeboe** [comment]
For us, imagestreams were working on 3.6. After stepping up to 3.7, we run in the same issue as discussed here... out image streams are trying to use IP iso SVC name of the registry.

On 3.6:

    apiVersion: v1
    kind: ImageStream
    metadata:
      annotations:
        openshift.io/image.dockerRepositoryCheck: '2017-12-11T13:56:46Z'
      creationTimestamp: '2017-12-11T13:56:46Z'
      generation: 2
      labels:
        app: my-image
      name: my-image
      namespace: test
      resourceVersion: '2133099'
      selfLink: /oapi/v1/namespaces/test/imagestreams/my-image
      uid: 207756ac-de7b-11e7-b435-0050568318f3
    spec:
      lookupPolicy:
        local: false
      tags:
        - annotations:
            description: Test image
            tags: my-image
          from:
            kind: DockerImage
            name: 'test/my-image:1.0.0'
          generation: 2
          importPolicy: {}
          name: 1.0.0
          referencePolicy:
            type: Source
    status:
      dockerImageRepository: 'docker-registry.default.svc:5000/test/my-image'

On 3.7:

    apiVersion: image.openshift.io/v1
    kind: ImageStream
    metadata:
      annotations:
        openshift.io/image.dockerRepositoryCheck: '2017-12-22T12:02:00Z'
      creationTimestamp: '2017-12-22T12:02:00Z'
      generation: 4
      labels:
        app: my-image
      name: my-image
      namespace: test
      resourceVersion: '569887'
      selfLink: >-
        /apis/image.openshift.io/v1/namespaces/test/imagestreams/my-image
      uid: eac5dcf7-e70f-11e7-ab97-00505683af24
    spec:
      lookupPolicy:
        local: false
      tags:
        - annotations:
            description: Test image
            tags: my-image
          from:
            kind: DockerImage
            name: 'test/my-image:1.0.0'
          generation: 4
          importPolicy: {}
          name: 1.0.0
          referencePolicy:
            type: Source
    status:
      dockerImageRepository: '10.2.85.87:5000/test/my-image'

**Comment by smarterclayton** [comment]
How are you configuring your registry hostname?  Check the
OPENSHIFT_DEFAULT_REGISTRY env var for your master processes and the
master-config.yaml section for external docker registry.  Is either of
those set to an IP address?

**Comment by boeboe** [comment]
@smarterclayton this variable is not set at all, not in any of the following locations:

- /etc/origin/master/master-config.yaml
- /etc/sysconfig/atomic-openshift-master-api
- /etc/sysconfig/atomic-openshift-master-controllers

Also the section mentioned as mentioned here (https://docs.openshift.com/container-platform/3.7/release_notes/ocp_3_7_release_notes.html#ocp-37-public-pull-url-provided-for-images) is not existing:

      imagePolicyConfig:
        internalRegistryHostname: docker-registry.default.svc:5000

**Comment by smarterclayton** [comment]
The hostname will *only* be used if you have set the config.  I would
recommend using the master-config.yaml setting and then using "oc adm
migrate image-references" to handle the change.

**Comment by sdodson** [comment]
@boeboe how'd you upgrade from 3.6 to 3.7? the upgrade playbooks shouldn't have removed OPENSHIFT_DEFAULT_REGISTRY from /etc/sysconfig/atomic-openshift-master-{api,controllers} If it did then we need to fix that. You anytime we update those files they should create a timestamped backup file you can check for the diff.

**Comment by boeboe** [comment]
The problem was solved using the following undocumented ansible variable:

    # registry options
    openshift_master_image_policy_config: '{"InternalRegistryHostname": "docker-registry.default.svc:5000"}'

**Comment by srinivas-coimbatore** [comment]
We are working with OpenShift 3.3, can you please tell us since which version is the OPENSHIFT_DEFAULT_REGISTRY variable supported?

**Comment by twforeman** [comment]
We upgraded to 3.7 recently and have this issue. It looks like there is a bug in the ansible upgrade playbook. The old timestamped files on our masters contain:

OPENSHIFT_DEFAULT_REGISTRY=docker-registry.default.svc:5000

The new versions do not.
